fdmi/plugins/fdmi_sample_plugin.c

```diff
@@ -0,0 +1,426 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+#define M0_TRACE_SUBSYSTEM M0_TRACE_SUBSYS_FDMI
+#include "lib/trace.h"
+
+#include <unistd.h>
+#include <fcntl.h>
+#include "fdmi/plugins/fdmi_sample_plugin.h"
+
+/**
+ * This is the FDMI plugin program. Its purpose is to start the listener
+ * for the special type of the FDMI events and forward them to the fdmi_app
+ * wrapper written in python. The communication is done via stdout.
+ */
+
+/**
+ * Storage for the program run parameters passed from console or the wrapper
+ * application.
+ */
+struct m0_fsp_params fsp_params;
+
+/**
+ * The main loop semaphore. The main loop here is just sleeping most of
+ * the time and waiting for some stop signal (e.g ctrl+c) from the python
+ * fdmi_app that controls this plugin program.
+ *
+ * The main job is done in one of the Motr locality threads, that handles
+ * the fops with FDMI data and call the callback function installed by this
+ * program if the type of the plugin match is found.
+ */
+static struct m0_semaphore fsp_sem;
+
+/**
+ * This program is basically adding the new client to the cluster. To do so
+ * we need to specify the client endpoint and the address of the cluster
+ * configuration service. This structure is used for delivering this info
+ * to the Motr core.
+ */
+static struct m0_config	fsp_conf = {};
+
+/**
+ * Pointer to the initialized m0_client structure denoting our program
+ * functionality.
+ */
+static struct m0_client	*fsp_client = NULL;
+
+/* DIX service (client interface for kv storage) config. */
+static struct m0_idx_dix_config	fsp_dix_conf = {};
+
+/**
+ * The vector of the FDMI plugin callback operations. They are called by
+ * the Motr when the new FDMI event occurs.
+ */
+const struct m0_fdmi_pd_ops *fsp_pdo;
+
+/**
+ * The handle of the local FDMI service that we start. Its purpose is to
+ * receive the FDMI rpcs, find the local consumer (this program) and send
+ * forward them to the consumer plugin in the form of FDMI records.
+ */
+static struct m0_reqh_service *fsp_fdmi_service = NULL;
+
+/**
+ * Aux buffer, used for data parsing.
+ */
+#define MAX_LEN 8192
+static char buffer[MAX_LEN];
+
+
+M0_INTERNAL char *to_hex(void *addr, int len)
+{
+	int i, j;
+	if ((2 * len) > MAX_LEN) {
+		fprintf(stderr, "to_hex() failed with (2 * len) > MAX_LEN\n");
+		return NULL;
+	}
+	for(i = 0, j = 0; i < (2 * len) && j < len; ++j)
+		i += sprintf(buffer + i, "%02x", ((char *)addr)[j]);
+	buffer[2 * len - 2] = '\0';
+	return buffer;
+}
+
+M0_INTERNAL void m0_dump_m0_fol_rec_to_json(struct m0_fol_rec *rec)
```

Comment:
Please change this function to "static".
Please remove the "m0_" prefix from its name.
"m0_" prefix usually is used for global and public APIs or variables.

Reply:
done